FIAT/polynomial_set.py

```diff
@@ -176,15 +176,51 @@ def polynomial_set_union_normalized(A, B):
     not contain any of the same members of the set, as we construct a
     span via SVD.
     """
-    new_coeffs = numpy.concatenate((A.coeffs, B.coeffs), axis=0)
+    assert A.get_reference_element() == B.get_reference_element()
+    new_coeffs = construct_new_coeffs(A.get_reference_element(), A, B)
+
+    deg = max(A.get_degree(), B.get_degree())
+    em_deg = max(A.get_embedded_degree(), B.get_embedded_degree())
     coeffs = spanning_basis(new_coeffs)
     return PolynomialSet(A.get_reference_element(),
-                         A.get_degree(),
-                         A.get_embedded_degree(),
+                         deg,
+                         em_deg,
                          A.get_expansion_set(),
                          coeffs)
 
 
+def construct_new_coeffs(ref_el, A, B):
+    # Constructs new coefficients for the set union of A and B
+    # If A and B do not have the same degree the smaller one
+    # is extended to match the larger
+
+    sd = ref_el.get_spatial_dimension()
+    if A.get_embedded_degree() != B.get_embedded_degree():
+        higher = A if A.degree > B.degree else B
+        lower = B if A.degree > B.degree else A
+
+        try:
+            sd = lower.get_shape()[0]
+        except IndexError:
+            sd = 1
+
+        embedded_coeffs = []
+        diff = higher.coeffs.shape[-1] - lower.coeffs.shape[-1]
+        for coeff in lower.coeffs:
+            if sd > 1:
+                new_coeff = []
+                for row in coeff:
+                    new_coeff.append(numpy.append(row, [0 for i in range(diff)]))
+                embedded_coeffs.append(new_coeff)
+            else:
+                embedded_coeffs.append(numpy.append(coeff, [0 for i in range(diff)]))
+        embedded_coeffs = numpy.array(embedded_coeffs)
+        new_coeffs = numpy.array(list(embedded_coeffs) + list(higher.coeffs))
+    else:
```

Comment:
this case works if A and B have the same continuity and degree.

Reply:
I guess then this function still needs to handle the case where A and B are not discontinuous but don't have the same degree. I haven't encountered this yet but will add it as a note in the function.